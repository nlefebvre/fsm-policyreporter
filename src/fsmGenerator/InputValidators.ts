enum FSMPart {
  STATE = "State",
  ALPHABET = "Alphabet",
  TRANSITION_FN = "Transition function"
}

export interface ErrorMsg {
  inputValue: string | Array<string> | null;
  type: FSMPart;
  message: string;
}


export const stateErrors = {
  duplicateEntries: "Duplicate States exist in state list",
  noEntries: "Must include at least one state"
}

const findDuplicate = (set: Array<string>): Array<string> => {
  const values: Array<string> = [];
  return set.reduce<Array<string>>((dupes, value) => {
    if (values.includes(value)) {
      dupes.push(value);
    } else {
      values.push(value);
    }
    return dupes;
  }, []);
}


export const validateStates = (stateArray?: Array<string>): Array<ErrorMsg> => {
  // empty array
  if (!stateArray || stateArray.length === 0) {
    return [
      {
        inputValue: null,
        type: FSMPart.STATE,
        message: stateErrors.noEntries
      }
    ]
  }

  // duplicate state
  // TODO ignorable error?
  const dupes = findDuplicate(stateArray);
  return dupes.map((dupe) => {
    return {
      inputValue: dupe,
      type: FSMPart.STATE,
      message: stateErrors.duplicateEntries
    }
  });
}

export const alphabetErrors = {
  noEntries: "Must include at east one alphabet",
  duplicateEntries: "Duplicate alphabet character exist in state list",

}

export const validateAlphabet = (alphabetArray: Array<string>): Array<ErrorMsg> => {
  // empty array
  if (!alphabetArray || alphabetArray.length === 0) {
    return [
      {
        inputValue: null,
        type: FSMPart.ALPHABET,
        message: alphabetErrors.noEntries
      }
    ]
  }

  // duplicate alphabet chars
  // TODO ignorable error?
  const dupes = findDuplicate(alphabetArray);

  return dupes.map((dupe) => {
    return {
      inputValue: dupe,
      type: FSMPart.ALPHABET,
      message: alphabetErrors.duplicateEntries
    };
  });

}


export const initialStateErrors = {
  invalidState: "Initial state does not exist on state list",
  noState: "Must include an initial state",
}

export const validateInitialState = (stateArray: Array<string>, initialState: string): Array<ErrorMsg> => {
  if (!initialState || initialState === "") {
    return [
      {
        inputValue: null,
        type: FSMPart.STATE,
        message: initialStateErrors.noState
      }
    ]
  }

  if (!stateArray.includes(initialState)) {
    return [
      {
        inputValue: initialState,
        type: FSMPart.STATE,
        message: initialStateErrors.invalidState
      }
    ]
  }
  return [];
}


export const finalStateErrors = {
  invalidState: "Final state(s) don't exist on state list",
  noState: "Must include a final state",
}

export const validateFinalStates = (stateArray: Array<string>, finalStateArray: Array<string>): Array<ErrorMsg> => {
  if (!finalStateArray || finalStateArray.length === 0) {
    return [
      {
        inputValue: null,
        type: FSMPart.STATE,
        message: finalStateErrors.noState
      }
    ]
  }
  const errors: Array<ErrorMsg> = [];

  finalStateArray.forEach((finalState) => {
    if (!stateArray.includes(finalState)) {
      errors.push({
        inputValue: finalState,
        type: FSMPart.STATE,
        message: finalStateErrors.invalidState
      });
    }
  })

  return errors;
}


export const transitionFnErrors = {
  noEntry: "No transitions found",
  missingStartingState: "Transition has invalid starting state",
  missingEndingState: "Transition has invalid ending state",
  missingTransition: "Transition value does not exist in alphabet"
};

export const validateTransitions = (stateArray: Array<string>, alphabet: Array<string>, transitions?: Array<[string, string, string]>): Array<ErrorMsg> => {
  const errors: Array<ErrorMsg> = [];

  if (!transitions) {
    return [
      {
        inputValue: null,
        type: FSMPart.STATE,
        message: transitionFnErrors.noEntry
      }
    ]
  }

  transitions.forEach(([startingState, transitionVal, endingState]) => {

    if (!stateArray.includes(startingState)) {
      errors.push({
        inputValue: startingState,
        type: FSMPart.STATE,
        message: transitionFnErrors.missingStartingState
      });
    }


    if (!stateArray.includes(endingState)) {
      errors.push({
        inputValue: endingState,
        type: FSMPart.STATE,
        message: transitionFnErrors.missingEndingState
      });
    }


    if (!alphabet.includes(transitionVal)) {
      errors.push({
        inputValue: transitionVal,
        type: FSMPart.STATE,
        message: transitionFnErrors.missingTransition
      });
    }
  })

  return errors;
}

export const validateAll = (json: Record<string, any>) => {
  const errors = []
  errors.push(...validateStates(json.states));
  errors.push(...validateAlphabet(json.alphabet));
  errors.push(...validateInitialState(json.states, json.initialState));
  errors.push(...validateFinalStates(json.states, json.finalStates));
  errors.push(...validateTransitions(json.states, json.alphabet, json.transitions));
  return errors;
}