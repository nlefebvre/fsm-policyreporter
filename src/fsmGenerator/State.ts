export class State {
  value: string;
  finalState: boolean;
  transition: Record<string, State | null>;

  constructor(value: string, finalState: boolean, transition: Record<string, State>) {
    this.value = value;
    this.finalState = finalState;
    this.transition = transition;
  }

  addTransition(alpha: string, state: State) {
    if (this.transition[alpha]) {
      throw Error(`Transition already exists!`)
    } else {
      this.transition[alpha] = state;
    }
  }

  removeTransition(alpha: string) {
    this.transition[alpha] = null;
  }

  setFinal(isFinal: boolean): void {
    this.finalState = isFinal;
  }

  isFinal() {
    return this.finalState;
  }

  process(alpha: string) {
    if (this.transition[alpha]) {
      return this.transition[alpha]
    }
    throw Error(`Missing transition for value ${alpha} on state '${this.value}`)
  }
}