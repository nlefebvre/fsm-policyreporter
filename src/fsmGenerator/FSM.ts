import { State } from "./State";

export class FSM {
  private states: Record<string, State>;
  private initialState: State;
  private currentState: State;

  constructor(
    states: Array<string>,
    initialState: string,
    finalStates: Array<string>,
    transitions: Array<[string, string, string]>) {
    this.states = states.reduce((states, stateValue: string) => {
      return { ...states, [stateValue]: new State(stateValue, false, {}) };
    }, {})
    this.initialState = this.states[initialState];
    this.currentState = this.initialState;

    finalStates.forEach((finalState) => {
      this.states[finalState].setFinal(true);
    });

    transitions.forEach(([initState, alpha, final]) => {
      this.states[initState].addTransition(alpha, this.states[final])
    })
  }

  getStateByValue(value: string) {
    return this.states[value];
  }

  getCurrentState() {
    return this.currentState;
  }

  private setCurrentState(state: State) {
    return this.currentState = state;
  }

  handleInput(input: string) {
    const inputArray = Array.from(input);
    try {
      inputArray.forEach((inputChar) => {
        this.setCurrentState(this.currentState.process(inputChar));
      });
    } catch (error) {
      console.error('error processing input', input);
    }

    return this.currentState;
  }

  reset() {
    this.currentState = this.initialState;
  }
}

