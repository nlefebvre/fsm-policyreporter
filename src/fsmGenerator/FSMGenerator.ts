import { FiniteStateMachine } from "../fsm/FiniteStateMachine";
import { validateAll } from "../validation/InputValidators";
import { State } from "../fsm/State";
import { generateErrorMessage } from "../validation/Errors";

export class FSMGenerator {
  states: Array<string>;
  initialState: string;
  finalStates: Array<string>;
  alphabet: Array<string>;
  transitions: Array<[string, string, string]>

  constructor() {
    this.states = [];
    this.initialState = '';
    this.finalStates = [];
    this.alphabet = [];
    this.transitions = [];
    return this;
  }

  addStates(states: Array<string>) {
    this.states.push(...states);
    return this;
  }

  setInitialState(initState: string) {
    this.initialState = initState;
    return this;
  }

  addFinalStates(finalStates: Array<string>) {
    this.finalStates.push(...finalStates);
    return this;
  }

  addAlphabet(alphabet: Array<string>) {
    this.alphabet.push(...alphabet);
    return this;
  }

  addTransitions(transitions: Array<[string, string, string]>) {
    this.transitions.push(...transitions);
    return this;
  }

  build() {
    const errors = validateAll({
      states: this.states,
      initialState: this.initialState,
      finalStates: this.finalStates,
      alphabet: this.alphabet,
      transitions: this.transitions
    });
    if (errors.length > 0) {
      throw new Error(generateErrorMessage("Error building Finite State Machine", errors));
    }

    const states = this.states.map((value) => {
      return new State(value, false, {});
    });
    const initialState: State = states.find((state) => this.initialState === state.value)!;
    const finiteStateMachine = new FiniteStateMachine(initialState);

    states.forEach((state) => {
      finiteStateMachine.addState(state);
    })

    this.finalStates.forEach((finalState) => {
      finiteStateMachine.getStateByValue(finalState).setFinal(true);
    });

    this.transitions.forEach(([startingStateValue, alphabet, endingStateValue]) => {
      const startingState = finiteStateMachine.getStateByValue(startingStateValue);
      const endingState = finiteStateMachine.getStateByValue(endingStateValue);
      startingState.addTransition(alphabet, endingState);
    });

    return finiteStateMachine;
  }
}