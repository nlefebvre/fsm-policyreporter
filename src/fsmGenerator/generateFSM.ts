import readline from 'readline';
import { readFile } from "fs/promises";
import { FSMGenerator } from './FSMGenerator';
import { ErrorMsg, validateAll, validateAlphabet, validateFinalStates, validateInitialState, validateStates, validateTransitions } from './InputValidators';


async function readJsonFile(path: string) {
  const file = await readFile(path, "utf8");
  return JSON.parse(file);
}

export const fromJSON = async (fileName: string) => {
  const jsonFile = await readJsonFile(fileName)
  const errors = validateAll(jsonFile);
  if (errors.length > 0) {
    console.error(generateErrorMessage("Error generating FSM from JSON", errors));
  } else {
    console.log(`Successfully loaded ${jsonFile.title ?? 'finite state machine'}`);
    const { states, initialState, finalStates, transitions } = jsonFile;
    return new FSMGenerator(states, initialState, finalStates, transitions as [string, string, string][]);
  }
}



export const fromInput = async () => {
  const q = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });

  const ask = (msg: string): Promise<string> => new Promise(resolve =>
    q.question(msg, response => resolve(response))
  );

  const states = await getInput(async () => {
    const statesInput = await ask("Input State Values (comma delimited):");
    const states = statesInput.split(",").filter((state) => state !== '');
    return states;
  }, validateStates);

  const alphabet = await getInput(async () => {
    const alphabetInput = await ask("Input Alphabet (each character will be included):");
    return alphabetInput.split("");
  }, validateAlphabet);

  const initialState = await getInput(async () => {
    return await ask("Input initial state. Must be values previously entered state list:");
  }, (initState: string) => {
    return validateInitialState(states, initState);
  });

  const finalStates = await getInput(async () => {
    const finalStateInput = await ask("Input final states (comma delimited). Must be values previously entered state list:");
    return finalStateInput.split(",").filter((state) => state !== '');;
  }, (finStates: Array<string>) => {
    return validateFinalStates(states, finStates);
  });


  const ingestTransition = async (transitions: Array<[string, string, string]>) => {
    const transitionInput = await ask("Input transition(comma delimited as startState, alphabet, endState). blank line once complete:");
    if (transitionInput) {
      const [start, alphabet, end] = transitionInput.split(",", 3)
      transitions.push([start, alphabet, end]);
      return ingestTransition(transitions);
    }
    return transitions;
  }

  const transitions = await getInput(async () => {
    return await ingestTransition([]);
  }, (trans: Array<[string, string, string]>) => {
    return validateTransitions(states, alphabet, trans)
  });

  q.close();
  return new FSMGenerator(states, initialState, finalStates, transitions)
}


const getInput = async <T>(inputFn: () => Promise<T>, validateFn: (input: T) => Array<ErrorMsg>) => {
  const input = await inputFn();
  const errs = validateFn(input);
  if (errs.length > 0) {
    console.log();
    console.error(generateErrorMessage("Error(s) with input", errs))
    console.log('Please re-enter.');
    return getInput(inputFn, validateFn);
  }
  return input;

}



const generateErrorMessage = (title: string, errors: Array<ErrorMsg>) => {
  return errors.reduce<string>((errorString, { type, message, inputValue }) => {
    return errorString += `    ${type}: ${message} ${inputValue !== null ? `(For value '${inputValue}')` : ''}\n`
  }, title + "\n")
}