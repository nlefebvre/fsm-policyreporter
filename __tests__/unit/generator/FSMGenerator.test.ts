import { FSMGenerator } from "../../../src/fsmGenerator/FSMGenerator"
import { mockAlphabet, mockFinalState, mockInitialState, mockStateArray, mockTransitions } from "../../data/mockData";

describe("FSMGenerator Tests", () => {

  test("FSMGenerator returns object", () => {
    expect(new FSMGenerator()).toBeDefined();
  });

  test("AddStates updates state", () => {
    const generator = new FSMGenerator();
    expect(generator.states).toStrictEqual([]);

    generator.addStates(mockStateArray);
    expect(generator.states).toStrictEqual(mockStateArray);
  });

  test("AddStates adds to existing state array", () => {
    const generator = new FSMGenerator()
      .addStates(mockStateArray)
      .addStates(mockStateArray);

    expect(generator.states).toStrictEqual([...mockStateArray, ...mockStateArray]);
  });

  test("addAlphabet updates alphabet", () => {
    const generator = new FSMGenerator();
    expect(generator.alphabet).toStrictEqual([]);

    generator.addAlphabet(mockAlphabet);
    expect(generator.alphabet).toStrictEqual(mockAlphabet);
  });

  test("addAlphabet adds to existing alphabet array", () => {
    const generator = new FSMGenerator()
      .addAlphabet(mockAlphabet)
      .addAlphabet(mockAlphabet);

    expect(generator.alphabet).toStrictEqual([...mockAlphabet, ...mockAlphabet]);
  });

  test("addFinalStates updates finalStates", () => {
    const generator = new FSMGenerator();
    expect(generator.finalStates).toStrictEqual([]);

    generator.addFinalStates(mockFinalState);
    expect(generator.finalStates).toStrictEqual(mockFinalState);
  });

  test("addFinalStates adds to existing finalStates array", () => {
    const generator = new FSMGenerator()
      .addFinalStates(mockFinalState)
      .addFinalStates(mockFinalState);

    expect(generator.finalStates).toStrictEqual([...mockFinalState, ...mockFinalState]);
  });


  test("setInitialState updates initialState", () => {
    const generator = new FSMGenerator();
    expect(generator.initialState).toBe('');

    generator.setInitialState(mockInitialState);
    expect(generator.initialState).toBe(mockInitialState);
  });

  test("addFinalStates adds to existing finalStates array", () => {
    const generator = new FSMGenerator()
      .addFinalStates(mockFinalState)
      .addFinalStates(mockFinalState);

    expect(generator.finalStates).toStrictEqual([...mockFinalState, ...mockFinalState]);
  });

  describe("build tests", () => {
    test("build returns object", () => {
      expect(new FSMGenerator()
        .addStates(mockStateArray)
        .addAlphabet(mockAlphabet)
        .addFinalStates(mockFinalState)
        .setInitialState(mockInitialState)
        .addTransitions(mockTransitions)
        .build()).toBeDefined();
    });

    test("build throws error if nothing is defined", () => {
      expect(() => new FSMGenerator().build()).toThrowErrorMatchingSnapshot();
    });

    test("build throws error for duplicate states", () => {
      const generator = new FSMGenerator()
        .addStates(mockStateArray)
        .addStates(mockStateArray)
        .addAlphabet(mockAlphabet)
        .addFinalStates(mockFinalState)
        .setInitialState(mockInitialState)
        .addTransitions(mockTransitions);

      expect(() => generator.build()).toThrowErrorMatchingSnapshot();
    });
  });
});

