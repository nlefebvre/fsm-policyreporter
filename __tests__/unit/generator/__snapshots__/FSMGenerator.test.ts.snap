// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FSMGenerator Tests build tests build throws error for duplicate states 1`] = `
"Error building Finite State Machine
    State: Duplicate States exist in state list (For value 'S0')
    State: Duplicate States exist in state list (For value 'S1')
    State: Duplicate States exist in state list (For value 'S2')
    State: Duplicate States exist in state list (For value 'S3')
    State: Duplicate States exist in state list (For value 'S4')
    State: Duplicate States exist in state list (For value 'S5')
    State: Duplicate States exist in state list (For value 'S6')
    State: Duplicate States exist in state list (For value 'S7')
    State: Duplicate States exist in state list (For value 'S8')
    State: Duplicate States exist in state list (For value 'S9')
    State: Duplicate States exist in state list (For value 'S10')
    State: Duplicate States exist in state list (For value 'S12')
    State: Duplicate States exist in state list (For value 'S13')
    State: Duplicate States exist in state list (For value 'S14')
    State: Duplicate States exist in state list (For value 'S15')
    State: Duplicate States exist in state list (For value 'S16')
"
`;

exports[`FSMGenerator Tests build tests build throws error if nothing is defined 1`] = `
"Error building Finite State Machine
    State: Must include at least one state 
    Alphabet: Must include at east one alphabet 
    State: Must include an initial state 
    State: Must include a final state 
"
`;
